### DISCUSSION TITLE
BindableState + SharedState strategy reseting binding

### DISCUSSION THREAD

**Discussion by derekc00** [pointfreeco/swift-composable-architecture#1661] [Q&A] [answered]
Hi, I'm attempting to use a BindableState property inside a state that is working on a subset of a parent state. I am experiencing a compiled file, but when I change the property's value, it resets back to its original value.

To provide more context, consider the [01-GettingStarted-SharedState case study](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-SharedState.swift). 

I added a simple BindableState Bool to the Profile State, added an action, BindingReducer, and toggle object. 
I placed the changes in [this gist](https://gist.github.com/derekc00/adc5ab9b80293be25aa0b6490da2b59d).

In this gif, I am repeatedly clicking the toggle button. This is the console output:

```received action:
  Root.Action.shared(
    SharedState.Action.profile(
      SharedState.Profile.Action.binding(
        BindingAction.set(
          WritableKeyPath<State, BindableState<Bool>>,
          true
        )
      )
    )
  )
  (No state changes)
```
However, I am expecting state changes to the `isOn` BindableState property.

![Simulator Screen Recording - iPhone 14 Pro - 2022-11-12 at 22 42 04](https://user-images.githubusercontent.com/43795676/201509339-36e3e26f-078d-48b6-815b-b04cc1d21b4e.gif)
![Screenshot 2022-11-12 at 10 29 22 PM](https://user-images.githubusercontent.com/43795676/201509451-37cd9f71-7bc5-440c-a2ef-54f9cb3676a9.png)

I'm 1 month into learning TCA and a few months into learning SwiftUI - And having a blast most of the time. I'm all ears to any hints or directions.

**Comment by tgrapperon** [comment, marked as answer]
Hey @derekc00, as you can see:
```swift
struct SharedState: ReducerProtocol {
  struct State: Equatable {
    var counter = Counter.State()
    var currentTab = Tab.counter

    var profile: Profile.State {
      get {
        Profile.State(
          currentTab: self.currentTab,
          count: self.counter.count,
          maxCount: self.counter.maxCount,
          minCount: self.counter.minCount,
          numberOfCounts: self.counter.numberOfCounts
        )
      }
      set {
        self.currentTab = newValue.currentTab
        self.counter.count = newValue.count
        self.counter.maxCount = newValue.maxCount
        self.counter.minCount = newValue.minCount
        self.counter.numberOfCounts = newValue.numberOfCounts
      }
    }
  }
  …
```
This means that the `Profile.State` value that is sent to your feature is a computed property. A new value will be derived each time, but you've not externalized the storage of your new flag, so the  updated `isOn` value is lost. One thing you can do is adding a private storage in `SharedState.State`:
```swift
struct SharedState: ReducerProtocol {
  struct State: Equatable {
    var counter = Counter.State()
    var currentTab = Tab.counter
+   var isOnStorage: Bool = false
    var profile: Profile.State {
      get {
        Profile.State(
          currentTab: self.currentTab,
          count: self.counter.count,
          maxCount: self.counter.maxCount,
          minCount: self.counter.minCount,
          numberOfCounts: self.counter.numberOfCounts,
+         isOn: self.isOnStorage
        )
      }
      set {
        self.currentTab = newValue.currentTab
        self.counter.count = newValue.count
        self.counter.maxCount = newValue.maxCount
        self.counter.minCount = newValue.minCount
        self.counter.numberOfCounts = newValue.numberOfCounts
+       self.isOnStorage = newValue.isOn
      }
    }
  }
  …
```
When there are many internal fields that you want to track, and when you're able to create a `State` value in absolute, without local parameters[^1], an alternative is to use a complete private `Profile.State` storage that you update instead of recreating a new value:
```swift
 struct State: Equatable {
    var counter = Counter.State()
    var currentTab = Tab.counter

    private var profileStorage = Profile.State()

    var profile: Profile.State {
      get {
        var state = profileStorage
        state.currentTab = self.currentTab
        state.count = self.counter.count
        state.maxCount = self.counter.maxCount
        state.minCount = self.counter.minCount
        state.numberOfCounts = self.counter.numberOfCounts
        return state
      }
      set {
        self.profileStorage = newValue
        // Synchronize other properties
        self.currentTab = newValue.currentTab
        self.counter.count = newValue.count
        self.counter.maxCount = newValue.maxCount
        self.counter.minCount = newValue.minCount
        self.counter.numberOfCounts = newValue.numberOfCounts
      }
    }
  }
  …
```
[^1]: It won't work here, as you can't create a `Profile.State()` like that, but I didn't want to file the missing arguments in the `init` with placeholder values to make the presentation clearer. Imagine that each property in `Profile.State` has a default value and that you can create `Profile.State()`, or that you create `var profileStorage = Profile.State(currentTab: .profile, count: 0, …` instead.

In both cases, the state of the `isOn` flag is persisted in the `SharedState.State` in some form or another, and your feature should work.

As an aside, I don't think that `SharedAction` needs to be `BindableAction` here. You're only using the `Profile` feature. You can probably remove all the `Binding`s stuff from the `SharedState` domain.

**Reply by derekc00** [reply]
Wow, thanks for a great explanation and quick reply, Thomas! I understand, now, how the computed property for the profile struct causes my binding to reset.

Cheers,
Derek
